boot/assets.go

```diff
@@ -289,14 +365,161 @@ type TrustedAssetsUpdateObserver struct{}
+		if err != nil {
+			return false, fmt.Errorf("cannot load modeenv: %v", err)
+		}
+	}
+	switch op {
+	case gadget.ContentUpdate:
+		return o.observeUpdate(whichBootloader, isRecovery, root, realSource, relativeTarget)
+	case gadget.ContentRollback:
+		return o.observeRollback(whichBootloader, isRecovery, root, realSource, relativeTarget)
+	default:
+		// we only care about update and rollback actions
+		return false, nil
+	}
+}
+
+func (o *TrustedAssetsUpdateObserver) observeUpdate(bl bootloader.Bootloader, recovery bool, root, realSource, relativeTarget string) (bool, error) {
+	modeenvBefore, err := o.modeenv.Copy()
+	if err != nil {
+		return false, fmt.Errorf("cannot copy modeenv: %v", err)
+	}
+
+	ta, err := o.cache.Add(realSource, bl.Name(), filepath.Base(relativeTarget))
+	if err != nil {
+		return false, err
+	}
+
+	trustedAssets := &o.modeenv.CurrentTrustedBootAssets
+	if recovery {
+		trustedAssets = &o.modeenv.CurrentTrustedRecoveryBootAssets
+	}
+	if !isAlreadyTrackedInBootMap(*trustedAssets, ta) {
+		if *trustedAssets == nil {
+			*trustedAssets = bootAssetsMap{}
+		}
+		(*trustedAssets)[ta.name] = append((*trustedAssets)[ta.name], ta.hash)
+	}
+
+	if o.modeenv.deepEqual(modeenvBefore) {
+		return true, nil
+	}
+	if err := o.modeenv.WriteTo(""); err != nil {
+		return false, fmt.Errorf("cannot write modeeenv: %v", err)
+	}
+	return true, nil
+}
+
+func (o *TrustedAssetsUpdateObserver) observeRollback(bl bootloader.Bootloader, recovery bool, root, realSource, relativeTarget string) (bool, error) {
 	// TODO:UC20:
-	// steps on write action:
-	// - copy new asset to assets cache
-	// - update modeeenv
 	// steps on rollback action:
 	// - drop file from cache if no longer referenced
 	// - update modeenv
-	return true, nil
+
+	trustedAssets := &o.modeenv.CurrentTrustedBootAssets
+	otherTrustedAssets := o.modeenv.CurrentTrustedRecoveryBootAssets
+	if recovery {
+		trustedAssets = &o.modeenv.CurrentTrustedRecoveryBootAssets
+		otherTrustedAssets = o.modeenv.CurrentTrustedBootAssets
+	}
+
+	assetName := filepath.Base(relativeTarget)
+	hashList, ok := (*trustedAssets)[assetName]
+	if !ok || len(hashList) == 0 {
+		// asset not tracked in modeenv
+		return true, nil
+	}
+
+	// new assets are appended to the list
+	expectedHash := hashList[0]
+	// sanity check, make sure that the current file is what we expect
+	newlyAdded := false
+	ondiskHash, err := o.cache.fileHash(filepath.Join(root, relativeTarget))
+	if err != nil {
+		// file may not exist if it was added by the update, that's ok
+		if !os.IsNotExist(err) {
+			return false, fmt.Errorf("cannot calculate the digest of current asset: %v", err)
+		}
+		newlyAdded = true
+	} else {
+		if ondiskHash != expectedHash {
+			// this is unexpected, a different file exists on disk?
+			return false, fmt.Errorf("unexpected content of existing asset %q", relativeTarget)
+		}
+	}
+
+	// XXX: do we need to support using the same asset name multiple times
+	// for a given bootloader?
```

Comment:
I would say no, can we error on the write path on that though?

Reply:
Yes, we can error out early in the backup pass, and similarly in the initial install observer.

Reply:
Added to both observers. I'll push the bit in initial install observer once #9160 lands.

Reply:
This is done in #9206 for the initial install observer.